### PR TITLE
[mlir] Optimize ThreadLocalCache by removing atomic bottleneck (attempt #3)

### PR DESCRIPTION
**Pull request by Mogball**
The ThreadLocalCache implementation is used by the MLIRContext (among other things) to try to manage thread contention in the StorageUniquers. There is a bunch of fancy shared pointer/weak pointer setups that basically keeps everything alive across threads at the right time, but a huge bottleneck is the `weak_ptr::lock` call inside the `::get` method.

This is because the `lock` method has to hit the atomic refcount several times, and this is bottlenecking performance across many threads. However, all this is doing is checking whether the storage is initialized. Importantly, when the `PerThreadInstance` goes out of scope, it does not remove all of its associated entries from the thread-local hash map (it contains dangling `PerThreadInstance *` keys). The `weak_ptr` also allows the thread local cache to synchronize with the `PerThreadInstance`'s destruction:

1. if `ThreadLocalCache` destructs, the `weak_ptr`s that reference its contained values are immediately invalidated
2. if `CacheType` destructs within a thread, any entries still live are removed from the owning `PerThreadInstance`, and it locks the `weak_ptr` first to ensure it's kept alive long enough for the removal.

This PR changes the TLC entries to contain a `shared_ptr<ValueT*>` and a `weak_ptr<PerInstanceState>`. It gives the `PerInstanceState` entries a `weak_ptr<ValueT*>` on top of the `unique_ptr<ValueT>`. This enables `ThreadLocalCache::get` to check if the value is initialized by dereferencing the `shared_ptr<ValueT*>` and check if the contained pointer is null. When `PerInstanceState` destructs, the values inside the TLC are written to nullptr. The TLC uses the
`weak_ptr<PerInstanceState>` to satisfy (2).

(1) is no longer the case. When `ThreadLocalCache` begins destruction, the `weak_ptr<PerInstanceState>` are invalidated, but not the `shared_ptr<ValueT*>`. This is OK: because the overall object is being destroyed, `::get` cannot get called and because the `shared_ptr<PerInstanceState>` finishes destruction before freeing the pointer, it cannot get reallocated to another `ThreadLocalCache` during destruction. I.e. the values inside the TLC associated with a `PerInstanceState` cannot be read during destruction. The most important thing is to make sure destruction of the TLC doesn't race with the destructor of `PerInstanceState`. Because `PerInstanceState` carries `weak_ptr` references into the TLC, we guarantee to not have any use-after-frees.

### CODE REVIEW

**Comment by Mogball** [comment]
This is a resubmit of the second version, which I hope fixes the underlying race issue. Someone please carefully read this code and make sure I'm right :( 

Also restores the removed header to ensure the build doesn't break.

**Comment by joker-eph** [comment]
I'd like to review this carefully, so please wait at least until next week before merging (I have plans for the long week end, not sure if I'll get to it)

**Comment by Mogball** [comment]
Enjoy your long weekend! ;)

File: mlir/include/mlir/Support/ThreadLocalCache.h [hunk 1]
```diff
@@ -25,28 +25,80 @@ namespace mlir {
 /// cache has very large lock contention.
 template <typename ValueT>
 class ThreadLocalCache {
+  struct PerInstanceState;
+
+  /// The "observer" is owned by a thread-local cache instance. It is
+  /// constructed the first time a `ThreadLocalCache` instance is accessed by a
+  /// thread, unless `perInstanceState` happens to get re-allocated to the same
+  /// address as a previous one. This class is destructed the thread in which
+  /// the `thread_local` cache lives is destroyed.
```
Comment (joker-eph) [inline mlir/include/mlir/Support/ThreadLocalCache.h:34]: Can you proof read this, it seems like it is missing something.

Comment (Mogball) [reply mlir/include/mlir/Support/ThreadLocalCache.h:34]: Yep. Thanks!

File: mlir/include/mlir/Support/ThreadLocalCache.h [hunk 2]
```diff
@@ -25,28 +25,80 @@ namespace mlir {
...
+  /// thread-local caches are owned by `PerInstanceState`, a reference is stored
+  /// via this class in the TLC. With a double pointer, it knows when the
+  /// referenced value has been destroyed.
+  struct Observer {
+    /// This is the double pointer, explicitly allocated because we need to keep
+    /// the address stable if the TLC map re-allocates. It is owned by the
+    /// observer and shared with the value owner.
+    std::shared_ptr<ValueT *> ptr = std::make_shared<ValueT *>(nullptr);
+    /// Because `Owner` living inside `PerInstanceState` contains a reference to
+    /// the double pointer, and livkewise this class contains a reference to the
```
Comment (jpienaar) [inline mlir/include/mlir/Support/ThreadLocalCache.h:46]: likewise

File: mlir/include/mlir/Support/ThreadLocalCache.h [hunk 3]
```diff
@@ -25,28 +25,80 @@ namespace mlir {
...
+  /// This class is called the "observer" because while values cached in
+  /// thread-local caches are owned by `PerInstanceState`, a reference is stored
+  /// via this class in the TLC. With a double pointer, it knows when the
+  /// referenced value has been destroyed.
+  struct Observer {
+    /// This is the double pointer, explicitly allocated because we need to keep
+    /// the address stable if the TLC map re-allocates. It is owned by the
+    /// observer and shared with the value owner.
+    std::shared_ptr<ValueT *> ptr = std::make_shared<ValueT *>(nullptr);
+    /// Because `Owner` living inside `PerInstanceState` contains a reference to
```
Comment (jpienaar) [inline mlir/include/mlir/Support/ThreadLocalCache.h:45]: I'd almost exchange the two sentences or add another one here.

**Review by jpienaar — COMMENTED** [review]
Were you able to repro the failures from previous attempts?

**Comment by Mogball** [comment]
> Were you able to repro the failures from previous attempts?

I was able to repro them and I confirmed this iteration fixes them. I also checked with TSAN.

**Pull request merged**